### PR TITLE
Update auto attach to allow for addons and role [ENT-718]

### PR DESCRIPTION
**Pull request by wottop**

### CODE REVIEW

File: server/src/main/resources/rules/rules.js [hunk 1]
```diff
@@ -1021,6 +1063,7 @@ var CoverageCalculator = {
              *        be added to handle that case.
              */
             default: function(complianceTracker, prodAttr, consumer) {
+                log.debug("CoverageCalc.default");
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Is this statement useful?

File: server/src/main/resources/rules/rules.js [hunk 2]
```diff
@@ -1311,8 +1354,29 @@ function createComplianceTracker(consumer, id) {
 
                 guest_limit: function (currentStackValue, poolValue, pool, quantity) {
                     return -1; //Value doesn't matter, just need it to be enforced
-                }
+                },
 
+                /**
+                 *  AddOns is accumulated by adding each pool value to
+                 *  a list of addon strings. Each pool value is a , separated
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Please change "," to "comma".  It took me a few seconds to figure out that a word wasn't missing here.

File: server/src/main/resources/rules/rules.js [hunk 3]
```diff
@@ -1311,8 +1354,29 @@ function createComplianceTracker(consumer, id) {
...
+                 */
+                addons: function (currentStackValue, poolValue, pool, quantity) {
+                    var stackValue = currentStackValue || [];
+                    stackValue.push(poolValue);
+                    return stackValue;
+                },
+
+                /**
+                 *  Role is accumulated by adding each pool value to
+                 *  a list of role strings. Each pool value is a , separated
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Same as the previous comment.

File: server/src/main/resources/rules/rules.js [hunk 4]
```diff
@@ -2291,7 +2355,15 @@ var Autobind = {
                         for (var i = 0; i < this.pools.length; i++) {
                             var pool = this.pools[i];
                             var prodAttrValue = pool.getProductAttribute(attr);
-                            if (!prodAttrValue || prodAttrValue === null) {
+                            var poolRole = pool.getProductAttribute("role");
+                            var poolAddons = pool.getProductAttribute("addons") != null?pool.getProductAttribute("addons").split(","):[];
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Fix the spacing here, please.

File: server/src/main/resources/rules/rules.js [hunk 5]
```diff
@@ -2351,9 +2423,18 @@ var Autobind = {
                 var provided_size = this.get_provided_products().length;
                 for (var i = this.pools.length - 1; i >= 0; i--) {
                     temp = this.pools[i];
+
+                    var roleMatch = role == temp.getProductAttribute("role");
+                    var poolAddons = temp.getProductAttribute("addons") != null ?
+                        temp.getProductAttribute("addons").split(","):[];
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: I'm not sure if JS's `split` takes a regex, but if it does we'll want to account for spaces here too.

File: server/src/main/resources/rules/rules.js [hunk 6]
```diff
@@ -2492,6 +2573,42 @@ var Autobind = {
...
+                var addons = [];
+                for (var i = 0; i < in_pools.length; i++) {
+                    var pool = in_pools[i];
+                    var pool_addons = pool.retrievePoolAttributeValues("addons");
+                    for (var k = 0; k < pool_addons.length; k++) {
+                        addons.push(pool_addons[k]);
+                    }
+                }
+                log.debug("get_addons_pools:  " + addons);
+                return addons;
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: This list could possibly have duplicates in it.  Is that something to worry about?

File: server/src/main/resources/rules/rules.js [hunk 7]
```diff
@@ -2734,8 +2851,41 @@ var Autobind = {
...
+     * Returns the list of productIds that the stack will cover, which the consumer requires.
+     */
+    get_common_role: function(role, group) {
+        if (role == null) {
+            return "";
+        }
+        var group_roles = group.get_roles();
+        for (var i = 0; i < group_roles.length; i++) {
+            var group_role = group_roles[i];
+            if (group_role == role) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `===`

File: server/src/main/resources/rules/rules.js [hunk 8]
```diff
@@ -2751,12 +2901,15 @@ var Autobind = {
             var group = all_groups[i];
             var group_avg_prio = group.get_average_priority();
             var intersection = this.get_common_products(installed, group).length;
+            var role_needed = this.get_common_role(role, group) != "";
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `!==`

File: server/src/main/resources/rules/rules.js [hunk 9]
```diff
@@ -2830,28 +2979,44 @@ var Autobind = {
...
+        var group = this.find_best_ent_group(all_groups, installed, role, addons);
         while (group != null) {
             best.push(group);
-            var in_common = this.get_common_products(installed, group);
+            var prods_in_common = this.get_common_products(installed, group);
             for (var j = installed.length - 1; j >= 0; j--) {
                 var current = installed[j];
-                if (in_common.indexOf(current) != -1) {
-                   installed.splice(j, 1);
+                if (prods_in_common.indexOf(current) != -1) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `!==`

File: server/src/main/resources/rules/rules.js [hunk 10]
```diff
@@ -2830,28 +2979,44 @@ var Autobind = {
...
+                    installed.splice(j, 1);
+                }
+            }
+            group.installed = prods_in_common;
+
+            var addons_in_common = this.get_common_addons(addons, group);
+            for (var j = addons.length -1; j >= 0; j--) {
+                var current = addons[j];
+                if (addons_in_common.indexOf(current) != -1) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `!==`

File: server/src/main/resources/rules/rules.js [hunk 11]
```diff
@@ -2830,28 +2979,44 @@ var Autobind = {
...
+                    addons.splice(j, 1);
                 }
             }
-            group.installed = in_common;
-            group = this.find_best_ent_group(all_groups, installed);
+            group.addons = addons_in_common;
+
+            var role_in_common = this.get_common_role(role, group);
+            if (role_in_common.indexOf(role) != -1) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `!==`

File: server/src/main/resources/rules/rules.js [hunk 12]
```diff
@@ -2886,6 +3051,47 @@ var Autobind = {
         return attached_ents;
     },
 
+    get_remaining_role: function (role, attached_ents) {
+        if (role == null) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `===`

File: server/src/main/resources/rules/rules.js [hunk 13]
```diff
@@ -2886,6 +3051,47 @@ var Autobind = {
...
+            return "";
+        }
+        if (attached_ents.length == 0 || role == "") {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `===`

File: server/src/main/resources/rules/rules.js [hunk 14]
```diff
@@ -2886,6 +3051,47 @@ var Autobind = {
...
+            return "";
+        }
+        else {
+            return role;
+        }
+    },
+
+
+    get_remaining_addons: function (addons, attached_ents) {
+        if (addons == null) {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `===`

File: server/src/main/resources/rules/rules.js [hunk 15]
```diff
@@ -2886,6 +3051,47 @@ var Autobind = {
...
+            return [];
+        }
+        var remaining_addons = [];
+        if (attached_ents.length == 0 || addons == "") {
```
Comment (awood) [inline server/src/main/resources/rules/rules.js]: Use `===`

**Comment by wottop** [comment]
retest this please

**Comment by wottop** [comment]
retest this please

**Comment by awood** [comment]
:moneybag: :gem: :champagne:

**Pull request merged**